plonky2/src/hash/hash_types.rs

```diff
@@ -83,27 +85,41 @@ where
 }
 
 impl<F: RichField> GenericHashOut<F> for HashOut<F> {
-    fn to_bytes(&self) -> Vec<u8> {
-        self.elements
-            .into_iter()
-            .flat_map(|x| x.to_canonical_u64().to_le_bytes())
-            .collect()
+    fn to_bytes(self) -> impl AsRef<[u8]> + AsMut<[u8]> + BorrowMut<[u8]> + Copy {
+        let mut bytes = [0u8; NUM_HASH_OUT_ELTS * 8];
+        for (i, x) in self.elements.into_iter().enumerate() {
+            let i = i * 8;
+            bytes[i..i + 8].copy_from_slice(&x.to_canonical_u64().to_le_bytes())
+        }
+        bytes
     }
 
     fn from_bytes(bytes: &[u8]) -> Self {
+        let mut bytes = bytes
+            .chunks(8)
+            .take(NUM_HASH_OUT_ELTS)
+            .map(|x| F::from_canonical_u64(u64::from_le_bytes(x.try_into().unwrap())));
+        HashOut {
+            elements: [(); NUM_HASH_OUT_ELTS].map(|()| bytes.next().unwrap()),
+        }
+    }
+
+    fn from_byte_iter(mut bytes: impl Iterator<Item = u8>) -> Self {
+        let bytes = [[(); 8]; NUM_HASH_OUT_ELTS].map(|b| b.map(|()| bytes.next().unwrap()));
+
+        HashOut {
+            elements: bytes.map(|x| F::from_canonical_u64(u64::from_le_bytes(x))),
+        }
+    }
+
+    fn from_iter(mut inputs: impl Iterator<Item = F>) -> Self {
         HashOut {
-            elements: bytes
-                .chunks(8)
-                .take(NUM_HASH_OUT_ELTS)
-                .map(|x| F::from_canonical_u64(u64::from_le_bytes(x.try_into().unwrap())))
-                .collect::<Vec<_>>()
-                .try_into()
-                .unwrap(),
+            elements: [(); NUM_HASH_OUT_ELTS].map(|()| inputs.next().unwrap()),
         }
     }
 
-    fn to_vec(&self) -> Vec<F> {
-        self.elements.to_vec()
+    fn into_iter(self) -> impl Iterator<Item = F> {
```

Comment:
Shouldn't this be a trait implementation?

Reply:
Not sure what you're asking. This is using "[return position impl Trait](https://doc.rust-lang.org/reference/types/impl-trait.html#abstract-return-types)" to enable unique return types between the different implementations without associated types.